release: support cross-compiling for arm64-apple-darwin

fanquake opened the pull request:
With a few minor changes, our macOS toolchain, and Guix, supports building for `arm64-apple-darwin` (Apple M1). Newer [`config.guess`](https://git.savannah.gnu.org/cgit/config.git/plain/config.guess) and [`config.sub`](https://git.savannah.gnu.org/cgit/config.git/plain/config.sub) are required (via autoconf 2.71), as support for recognising `arm64-apple-darwin` [was only recently added](https://git.savannah.gnu.org/cgit/config.git/commit/?id=2593751ef276497e312d7c4ce7fd049614c7bf80). A newer version of the macOS SDK (11.0+) is also required; I've taken this from Xcode 12.2.

Guix Build:
```bash
bash-5.1# find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
c9c64c6252927c82d4a96eb2b1a13c557ac230afa3640a7f2feec87fabe2b7e5  guix-build-edd70b52fcd7/output/aarch64-linux-gnu/SHA256SUMS.part
b57dd461ec6a0b22ce9614ab824bcbff4c3e6e935c22367aff9a718f46c3fd39  guix-build-edd70b52fcd7/output/aarch64-linux-gnu/bitcoin-edd70b52fcd7-aarch64-linux-gnu-debug.tar.gz
66da9f57a6789108db3460d2b54a8bed68b7e147627404e51f66254d91500402  guix-build-edd70b52fcd7/output/aarch64-linux-gnu/bitcoin-edd70b52fcd7-aarch64-linux-gnu.tar.gz
13213523984837fc66fe27cc4dac269f6ef554ec56316af92337f4c4845f6acf  guix-build-edd70b52fcd7/output/arm-linux-gnueabihf/SHA256SUMS.part
e1a17b41c92879ccb6608a6a6f82aac22d999f10d798f40f96297d91ebd90d20  guix-build-edd70b52fcd7/output/arm-linux-gnueabihf/bitcoin-edd70b52fcd7-arm-linux-gnueabihf-debug.tar.gz
6a6d54081c2dd8f9a44293275aea875587b5fa7cc31a30925fb5156bb1e55bf3  guix-build-edd70b52fcd7/output/arm-linux-gnueabihf/bitcoin-edd70b52fcd7-arm-linux-gnueabihf.tar.gz
79f617a2f0a11e6a8fcf8058e05204a374ab5f689ccb15a2ff449ee2dd5e8216  guix-build-edd70b52fcd7/output/arm64-apple-darwin/SHA256SUMS.part
6fb04a18b17812174ec35a6788dce5745d6c2dec558056dcfdd356b57d089b99  guix-build-edd70b52fcd7/output/arm64-apple-darwin/bitcoin-edd70b52fcd7-arm64-apple-darwin.tar.gz
a67e415631bea940adab86adec9899593fa35e1773d4ac1149efd249adf89996  guix-build-edd70b52fcd7/output/arm64-apple-darwin/bitcoin-edd70b52fcd7-osx-unsigned.dmg
2a41c273afb8bb621226de80302a7190eee057c2447d113999f40a9485430d44  guix-build-edd70b52fcd7/output/arm64-apple-darwin/bitcoin-edd70b52fcd7-osx-unsigned.tar.gz
c8d9358471e1d72d56122ae6f53dc2a42ae0925628bf347acd1a3ae7110d8759  guix-build-edd70b52fcd7/output/dist-archive/bitcoin-edd70b52fcd7.tar.gz
1985a4869307418ec2ee85f5f0ccfea0944eba0792318a8676ae213a9589ed97  guix-build-edd70b52fcd7/output/powerpc64-linux-gnu/SHA256SUMS.part
6a1d3525011da84e24997c9ab39563b3597ae881b4d1af5a09a6e673cfc59808  guix-build-edd70b52fcd7/output/powerpc64-linux-gnu/bitcoin-edd70b52fcd7-powerpc64-linux-gnu-debug.tar.gz
cda50c28b2920caef74de4b3e78fd4d81592d1aadc9758d1882f1cad16dc83ca  guix-build-edd70b52fcd7/output/powerpc64-linux-gnu/bitcoin-edd70b52fcd7-powerpc64-linux-gnu.tar.gz
ca92936cf796543d13250f94175af772a9fc985e834017d6af2cfb36eea289e6  guix-build-edd70b52fcd7/output/powerpc64le-linux-gnu/SHA256SUMS.part
64c46248065149a384c5710adeb4fd8e795ffee7d6441e4edb71468c6cf1478e  guix-build-edd70b52fcd7/output/powerpc64le-linux-gnu/bitcoin-edd70b52fcd7-powerpc64le-linux-gnu-debug.tar.gz
bb9c3400a1dd88fa5fd26e8e22ba0f3a43263bdbb762233ed1670394d3d69ed1  guix-build-edd70b52fcd7/output/powerpc64le-linux-gnu/bitcoin-edd70b52fcd7-powerpc64le-linux-gnu.tar.gz
926ef7a23ecdf02a6619f9dba0db814373c1a5105347e4c13b80a116910d1c8d  guix-build-edd70b52fcd7/output/riscv64-linux-gnu/SHA256SUMS.part
cdf7ce78f206378a44122865150784c20bf1bec6943b48e35a41633c0f1f05ed  guix-build-edd70b52fcd7/output/riscv64-linux-gnu/bitcoin-edd70b52fcd7-riscv64-linux-gnu-debug.tar.gz
026bbb6662f4074054457ce1678fc1d697bfbf87c80f1c3d5d560bcc18e3bea6  guix-build-edd70b52fcd7/output/riscv64-linux-gnu/bitcoin-edd70b52fcd7-riscv64-linux-gnu.tar.gz
0190b93fc5d5202c6b252e51c34799fb153a284cc53b0e6f62d2521dc61866bd  guix-build-edd70b52fcd7/output/x86_64-apple-darwin/SHA256SUMS.part
69aac1a962aeeec01c30015f79d93c57d68a484c3905b8a526f90e7e7382bc1b  guix-build-edd70b52fcd7/output/x86_64-apple-darwin/bitcoin-edd70b52fcd7-osx-unsigned.dmg
ea37691f1df0389afedbb9eece45660dcb2ee867ffda1d26b244a471af2a6230  guix-build-edd70b52fcd7/output/x86_64-apple-darwin/bitcoin-edd70b52fcd7-osx-unsigned.tar.gz
f56ff4035e3bc70c2cbb04f8a7f5ef23096088189e84fa172cec97f59336cf2b  guix-build-edd70b52fcd7/output/x86_64-apple-darwin/bitcoin-edd70b52fcd7-osx64.tar.gz
a1c1bf47083c6f19bcc68761ea56bec666cfe128f4ba878a5655ae12874fe038  guix-build-edd70b52fcd7/output/x86_64-linux-gnu/SHA256SUMS.part
cad26c66c764361b30c8204e7743dbb6c0e6dd1eb3af9418d77bd566e05e1470  guix-build-edd70b52fcd7/output/x86_64-linux-gnu/bitcoin-edd70b52fcd7-x86_64-linux-gnu-debug.tar.gz
ed2b459a06d5bc0f895402323fc5cae543a387d0621a55214feb37513210324e  guix-build-edd70b52fcd7/output/x86_64-linux-gnu/bitcoin-edd70b52fcd7-x86_64-linux-gnu.tar.gz
```

jarolrod committed:
doc: add arm macOS depends platform triplet

fanquake committed:
build: use macOS 11 SDK (Xcode 12.2)

This should be sufficient to support building for Apple ARM when
cross-compiling.

promag commented:
Tried your binaries but they are faulty
```sh
bash-3.2$ ./bitcoind --help
Killed: 9
```
and
```sh
bash-3.2$ otool -L ./bitcoind
./bitcoind:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
```
But I don't have these libraries. Searching around gives https://developer.apple.com/forums/thread/655588?answerId=623718022#623718022 and https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-release-notes:

> New in macOS Big Sur 11.0.1, the system ships with a built-in dynamic linker cache of all system-provided libraries. As part of this change, copies of dynamic libraries are no longer present on the filesystem. Code that attempts to check for dynamic library presence by looking for a file at a path or enumerating a directory will fail. Instead, check for library presence by attempting to dlopen() the path, which will correctly check for the library in the cache. (62986286)

promag commented:
Did the same `otool -L src/bitcoind` on a binary built on m1 and have the same output
```
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
```
So I guess it's a different issue. Any hint?

fanquake commented:
@promag sorry for such a slow follow up. I've only recently gotten an M1 box, and am now getting back to this. As far as I'm aware, the issue you're seeing is related [to code-signing requirements](https://github.com/bitcoin/bitcoin/issues/22403#issuecomment-876873551). As suggested in #22403 if you run:
```bash
codesign --sign - --deep --force bitcoind
```

on my binaries they should then work. I have tested this after downloading them onto an M1 box, and running the ad-hoc signing. That's great, as it means our current toolchain seems to be setup to produce working M1 binaries (with the changes here).

I will figure out how best to incorporate that signing into the build process for non-release builds.

fanquake committed:
guix: add arm64-apple-darwin triplet

fanquake commented:
This is now based on #22436. I've added a commit to add the `arm64-apple-darwin20` host for Guix, fixed up the libevent change to account for #21991, fixed the failing macOS symbol check (needed to account for the new SDK) and rebased on master.

Running Guix builds I've found that we'll need to update our security-checks tests, because `ld64` ignores `-no_pie` for arm64:
```bash
ld: warning: -no_pie ignored for arm64
F
======================================================================
FAIL: test_MACHO (__main__.TestSecurityChecks)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./contrib/devtools/test-security-check.py", line 83, in test_MACHO
    self.assertEqual(call_security_check(cc, source, executable, ['-Wl,-no_pie','-Wl,-flat_namespace','-Wl,-allow_stack_execute','-fno-stack-protector']),
AssertionError: Tuples differ: (1, 'test1: failed NOUNDEFS NX LAZY_BINDINGS Canary CONTROL_FLOW') != (1, 'test1: failed PIE NOUNDEFS NX LAZY_BINDINGS Canary CONTROL_FLOW')

First differing element 1:
'test1: failed NOUNDEFS NX LAZY_BINDINGS Canary CONTROL_FLOW'
'test1: failed PIE NOUNDEFS NX LAZY_BINDINGS Canary CONTROL_FLOW'

- (1, 'test1: failed NOUNDEFS NX LAZY_BINDINGS Canary CONTROL_FLOW')
+ (1, 'test1: failed PIE NOUNDEFS NX LAZY_BINDINGS Canary CONTROL_FLOW')
?                   ++++
```

We also need to make some update in regards to how we are doing code signing. Outside of that, this PR currently produces working M1 binaries.

hebasto commented on depends/README.md:
```diff
@@ -28,7 +28,8 @@ Common `host-platform-triplet`s for cross compilation are:
 - `x86_64-w64-mingw32` for Win64
-- `x86_64-apple-darwin18` for macOS
+- `x86_64-apple-darwin18` for Intel macOS
+- `arm64-apple-darwin20` for ARM macOS
```
Why `arm64-apple-darwin20` if `depends/config.sub arm64-apple-darwin20` returns `aarch64-apple-darwin20`?

hebasto replied:
On second thought, I think `arm64-apple-darwin20` is the correct triplet. It corresponds with the `uname -m` output.

hebasto reviewed (commented):
Tested 53239569e2d2d041211ac37f6507ecbfb572fe00 on Linux Mint 20.2 (x86_64):
```
$ make -C depends qt_configured HOST=arm64-apple-darwin20
...
Configure summary:

Building on: linux-g++ (x86_64, CPU features: mmx sse sse2)
Building for: macx-clang-linux (x86_64, CPU features: cx16 mmx sse sse2 sse3 ssse3 sse4.1)
Target compiler: clang (Apple) 10.0.1
Configuration: cross_compile sse2 aesni sse3 ssse3 sse4_1 sse4_2 avx avx2 avx512f avx512bw avx512cd avx512dq avx512er avx512ifma avx512pf avx512vbmi avx512vl f16c largefile precompile_header rdrnd shani silent x86SimdAlways release c++11 c++14 c++1z reduce_exports static stl
```

I think in `Building for: ` should be `arm64` rather `x86_64`.

UPDATE: solved in https://github.com/bitcoin/bitcoin/pull/21851#pullrequestreview-813823129

hebasto reviewed (commented):
Drop outdated `contrib/gitian-build.py` and `contrib/gitian-descriptors/gitian-osx.yml`?

hebasto reviewed (commented):
> - [ ] Qt doesn't work yet. Unclear if it will be trivial to make work with 5.12.x.

The following patch should fix Qt package:

```diff
diff --git a/depends/packages/qt.mk b/depends/packages/qt.mk
index 12e0494ad..5bfd4f69c 100644
--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -127,7 +127,7 @@ $(package)_config_opts_darwin += -device-option MAC_TARGET=$(host)
 $(package)_config_opts_darwin += -device-option XCODE_VERSION=$(XCODE_VERSION)
 endif
 
-# for macOS on Apple Silicon (ARM) see https://bugreports.qt.io/browse/QTBUG-85279
+$(package)_config_opts_x86_64_darwin += -device-option QMAKE_APPLE_DEVICE_ARCHS=x86_64
 $(package)_config_opts_aarch64_darwin += -device-option QMAKE_APPLE_DEVICE_ARCHS=arm64
 
 $(package)_config_opts_linux = -qt-xcb
diff --git a/depends/patches/qt/mac-qmake.conf b/depends/patches/qt/mac-qmake.conf
index 190ab7a16..e4bfaa146 100644
--- a/depends/patches/qt/mac-qmake.conf
+++ b/depends/patches/qt/mac-qmake.conf
@@ -13,7 +13,6 @@ QMAKE_MAC_SDK.macosx.Path = $${MAC_SDK_PATH}
 QMAKE_MAC_SDK.macosx.platform_name = macosx
 QMAKE_MAC_SDK.macosx.SDKVersion = $${MAC_SDK_VERSION}
 QMAKE_MAC_SDK.macosx.PlatformPath = /phony
-QMAKE_APPLE_DEVICE_ARCHS=x86_64
 !host_build: QMAKE_CFLAGS += -target $${MAC_TARGET}
 !host_build: QMAKE_OBJECTIVE_CFLAGS += $$QMAKE_CFLAGS
 !host_build: QMAKE_CXXFLAGS += $$QMAKE_CFLAGS
```

---

FWIW, with the suggested changes the `bitcoind` and `bitcoin-qt` binaries being signed are fully functional.

hebasto commented on autogen.sh:
```diff
@@ -14,3 +14,15 @@ fi
+# autoreconf / libtoolize may generate versions of config.guess and config.sub
+# that are too old to recognise some triplets, i.e arm64-apple-darwin20, so
+# copy over the generated config.guess & config.sub with updated versions.
+cp depends/config.guess build-aux/config.guess
```
During Guix build:
```
cp: cannot create regular file 'build-aux/config.guess': Permission denied
```

cc @dongcarl

hebasto commented:
Restarted `macOS 10.15 [gui, no tests] [focal]` CI task as new SDK should be available now in https://bitcoincore.org/depends-sources/sdks/Xcode-12.2-12B45b-extracted-SDK-with-libcxx-headers.tar.gz

hebasto commented:
Tested 9ea2ec035dab9dff948b27e71703305d6adc5188: the [issue](https://github.com/bitcoin/bitcoin/pull/21851#discussion_r755348943) still persist.

hebasto commented:
Our current CI scripts are needed to be updated (see #23597) to handle macOS SDK version bump properly.

Here are:
- this branch based on the main commit of #23597 
- CI [log](https://cirrus-ci.com/task/6184831487836160) -- "macOS 10.15 [gui, no tests] [focal]" is GREEN :tiger2:

fanquake commented:
> Tested 9ea2ec0: the issue with Guix builds still persists.

The changes are still the same, so I wouldn't expect any different. This PR is marked as a draft; there are a number of things left to address. The Guix build wouldn't work regardless, because of the issues outlined in the PR description.

fanquake committed:
contrib: support arm64 darwin in security checks

fanquake committed:
guix: use autoconf 2.71

This allows for building with newer targets, like arm64-apple-darwin, due to
having a newer bundled config.guess and config.sub.

fanquake commented:
This is now ready for review. I've updated the security checks to account for arm64 Darwin. Switching to using autoconf-2.71 has solved the problems with outdated config.guess & config.sub. There will still be a few followup changes in regards to code-signing etc, but I think as-is, this change could go in, meaning we could be releasing 23.0  with arm64 Darwin binaries. I've added Guix hashes to the PR description.

hebasto commented:
#### Guix build:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
759c65a3c5477697f973b71e694f5876dacd760349be6ad0b24a5bbeaccaa86d  guix-build-f26d41ba8cbb/output/arm64-apple-darwin/SHA256SUMS.part
7c4899053cb53e04f333d39ecc42d3b7bdadc0310b04e1696ee59d4378ff279c  guix-build-f26d41ba8cbb/output/arm64-apple-darwin/bitcoin-f26d41ba8cbb-arm64-apple-darwin.tar.gz
38808a8fe9a872fa0b0a75e07a1784a9bf0e95d6e8b376f2fcd64ceb121cef3f  guix-build-f26d41ba8cbb/output/arm64-apple-darwin/bitcoin-f26d41ba8cbb-osx-unsigned.dmg
a09e894b2016626c0475594c1a2a033ce796fcbb45a30123d67cfdae805db595  guix-build-f26d41ba8cbb/output/arm64-apple-darwin/bitcoin-f26d41ba8cbb-osx-unsigned.tar.gz
5d230402f029280a33226c16f2bb680611f73478650fd0c61a919add72bbad29  guix-build-f26d41ba8cbb/output/dist-archive/bitcoin-f26d41ba8cbb.tar.gz
```

hebasto commented:
Tested on Guix build `bitcoin-edd70b52fcd7-osx-unsigned.dmg` on M1:

![Screenshot 2022-01-26 at 13 06 06](https://user-images.githubusercontent.com/32963518/151152692-ea20c681-bfcf-45c4-bf69-19c47cbcfe79.png)

And it seems not related to the lack of notarization.

fanquake commented:
Fixed the CI, and added hashes for all builds to the PR description.

> And it seems not related to the lack of notarization.

Can you try running the binary directly, and/or providing a more informative error message.

hebasto commented:
#### Guix builds:

```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
79f617a2f0a11e6a8fcf8058e05204a374ab5f689ccb15a2ff449ee2dd5e8216  guix-build-edd70b52fcd7/output/arm64-apple-darwin/SHA256SUMS.part
6fb04a18b17812174ec35a6788dce5745d6c2dec558056dcfdd356b57d089b99  guix-build-edd70b52fcd7/output/arm64-apple-darwin/bitcoin-edd70b52fcd7-arm64-apple-darwin.tar.gz
a67e415631bea940adab86adec9899593fa35e1773d4ac1149efd249adf89996  guix-build-edd70b52fcd7/output/arm64-apple-darwin/bitcoin-edd70b52fcd7-osx-unsigned.dmg
2a41c273afb8bb621226de80302a7190eee057c2447d113999f40a9485430d44  guix-build-edd70b52fcd7/output/arm64-apple-darwin/bitcoin-edd70b52fcd7-osx-unsigned.tar.gz
c8d9358471e1d72d56122ae6f53dc2a42ae0925628bf347acd1a3ae7110d8759  guix-build-edd70b52fcd7/output/dist-archive/bitcoin-edd70b52fcd7.tar.gz
0190b93fc5d5202c6b252e51c34799fb153a284cc53b0e6f62d2521dc61866bd  guix-build-edd70b52fcd7/output/x86_64-apple-darwin/SHA256SUMS.part
69aac1a962aeeec01c30015f79d93c57d68a484c3905b8a526f90e7e7382bc1b  guix-build-edd70b52fcd7/output/x86_64-apple-darwin/bitcoin-edd70b52fcd7-osx-unsigned.dmg
ea37691f1df0389afedbb9eece45660dcb2ee867ffda1d26b244a471af2a6230  guix-build-edd70b52fcd7/output/x86_64-apple-darwin/bitcoin-edd70b52fcd7-osx-unsigned.tar.gz
f56ff4035e3bc70c2cbb04f8a7f5ef23096088189e84fa172cec97f59336cf2b  guix-build-edd70b52fcd7/output/x86_64-apple-darwin/bitcoin-edd70b52fcd7-osx64.tar.gz
```

hebasto commented:
> Can you try running the binary directly, and/or providing a more informative error message.

For now I got
```
hebasto@mini ~ % Downloads/bitcoin-edd70b52fcd7/bin/bitcoind
zsh: killed     Downloads/bitcoin-edd70b52fcd7/bin/bitcoind
```

Trying to get more info.

hebasto commented:
> Fixed the CI, and added hashes for all builds to the PR description.
> 
> > And it seems not related to the lack of notarization.
> 
> Can you try running the binary directly, and/or providing a more informative error message.

#22546 ?

hebasto commented:
```
% codesign --sign - Downloads/bitcoin-edd70b52fcd7/bin/bitcoind
```
helps.

hebasto reviewed (approved):
ACK edd70b52fcd79026e9c63b8bc481fa68f95edf11, a Guix `bitcoin-edd70b52fcd7-osx-unsigned.dmg` tested on Mac mini (M1, 2020) + macOS Monterey 12.1 (21C52).

![Screenshot 2022-01-26 at 13 40 05](https://user-images.githubusercontent.com/32963518/151157091-3dbcd224-c779-4273-9caa-1d90d808ef36.png)
